### PR TITLE
fix(coding-agent): keep selection highlight visible against matching terminal backgrounds

### PR DESCRIPTION
**Pull request by xeophon**
## Problem

Selected rows (agents view, child agent panel, menu panel, tree selector, session selector) use a fixed `selectedBg` theme color. When the terminal background is close to that color, the selection highlight is nearly invisible. Concrete case: the `prime` theme ships `selectedBg: #222226` (luminance ~34), which against a Ghostty Gruvbox Dark Hard background `#1d2021` (luminance ~31) has a contrast delta of ~3.

The existing surface adaptation (minimum luminance delta for editor/popup backgrounds) is not applied to selection rows, and its delta of 12 is too weak for a selection indicator anyway.

## Fix

`Theme.getSelectionBackgroundColor()`: when the terminal background is known (OSC 10/11 probe) and the luminance delta against `selectedBg` is below 28, the highlight is blended toward white (or black on light terminals) by just enough alpha (capped at 0.5) to reach that minimum. Falls back to the plain theme color when the background can't be detected (e.g. tmux) or the delta is already sufficient, so well-contrasted setups render unchanged.

Adapted highlights use the detected terminal palette, consistent with how `getEditorBackgroundColor()`/`getPopupBackgroundColor()` already adapt passive surfaces.

## Testing

- New cases in `theme-adaptive.test.ts`: unknown background, sufficient contrast, dark terminal bg matching `selectedBg` (adapted), light terminal bg matching light-theme `selectedBg` (adapted).
- Existing selector/agent-list tests still pass.
- Verified against the installed 0.3.3 bundle: with a `#1d2021` terminal background the selection row renders `rgb(59,59,62)` instead of `rgb(34,34,38)`; unchanged for black or unknown backgrounds.

---

> [!NOTE]
> **Low Risk**
> TUI-only selection styling with fallbacks to the original theme token; no auth, data, or runtime behavior changes.
> 
> **Overview**
> Fixes **nearly invisible list selection** when the terminal background is close to the theme's `selectedBg` (e.g. prime `#222226` on Gruvbox `#1d2021`).
> 
> Adds **`Theme.getSelectionBackgroundColor()`**, which compares luminance against the detected terminal background (OSC probe). If contrast is below a **28** luminance delta, it blends toward white or black (capped alpha, with fallbacks when the same-side blend cannot reach the threshold) and accounts for **256-color quantization** so palette rounding does not erase the highlight. When the background is unknown or `selectedBg` is a basic ANSI index, behavior stays **`selectedBg` unchanged**.
> 
> **Agents view**, **child agent panel**, **menu panel**, and **tree selector** now use this helper instead of a fixed `selectedBg` background. Changelog and **`theme-adaptive.test.ts`** cover contrast, light/dark terminals, empty `selectedBg`, and quantization edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit efa39d1c9dcf013acef87a1d953bf41fc915eb0c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Fix selection highlight visibility against matching terminal backgrounds in coding agent UI
> - Adds `Theme.getSelectionBackgroundColor()` in [theme.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/562/files#diff-cfeb6df930e2178ee2bad20d06795540093a317adcd9ed03e9b70bf39c667c53), which returns an adaptive background function that blends the configured `selectedBg` toward black or white until a minimum luminance delta versus the terminal background is met.
> - Replaces direct `theme.bg("selectedBg", ...)` calls in [agents-view-mode.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/562/files#diff-116f4364505fa125d57532296cde69f9a73543363de3d3cd48bd4f48b2ee6e10), [child-agent-inspector.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/562/files#diff-c6e80ba5fe28e36eaaf2f0d66307dec3db0481b56d24b214c31d2530d49517a1), [menu-panel.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/562/files#diff-22f82f8366fad943dc6e7f231f4645c864e17b89b5228a1453cc21e1a57a329c), and [tree-selector.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/562/files#diff-eb3a2b5ed9743a3bf5f855316bb3f508141b0d28028ed4cc246be17d04a77fa6) with the new adaptive method.
> - Falls back to the configured `selectedBg` unchanged when the terminal background is unknown or when `selectedBg` is a basic ANSI color (indices 0–15).
> - Behavioral Change: selection highlight color may now differ from the configured `selectedBg` in truecolor and 256-color terminals when the background is too similar to the selection color.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized efa39d1.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 1]
```diff
@@ -420,6 +423,32 @@ export class Theme {
 		return this.surfaceBackgroundColor("toolPanelBg");
 	}
 
+	getSelectionBackgroundColor(): (str: string) => string {
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Correct — when the selection color is darker than the terminal background, blending toward white moved it toward the background and reduced the delta below its previous value. Fixed in ca19c7ca: the blend direction is now chosen by which side of the terminal background the selection sits on (brighter toward white, darker toward black). Added tests covering both sidedness cases.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 2]
```diff
@@ -420,6 +423,32 @@ export class Theme {
...
+		if (delta >= SELECTION_MIN_LUMINANCE_DELTA) {
+			return (str: string) => this.bg("selectedBg", str);
+		}
+
+		const top = isLightColor(terminalBg) ? BLACK : WHITE;
+		const topDelta = Math.abs(luminance(top) - luminance(selectionRgb));
+		const alpha =
+			topDelta > 0
+				? Math.min(SELECTION_MAX_BLEND_ALPHA, (SELECTION_MIN_LUMINANCE_DELTA - delta) / topDelta)
+				: SELECTION_MAX_BLEND_ALPHA;
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in a3506b00 — same root cause as macroscope's comment above; the blend now picks the endpoint on the selection's side of the terminal background, with a cross-side fallback.

File: packages/coding-agent/CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,8 @@
 - Changed session search to rank exact name and first-message matches before prefix, substring, and stricter transcript fuzzy matches.
 - Fixed daemon startup failing when an interrupted supervisor left an empty ownership directory.
 - Fixed `/effort xhigh` and `/effort max` being rejected as unknown levels when no model was active yet.
+- Fixed selected rows in the agents view, child agent panel, and selectors being nearly invisible on terminals whose background closely matches the `selectedBg` theme color; the selection highlight now adapts to keep a minimum contrast against the detected terminal background.
+ (fix(coding-agent): keep selection highlight visible against matching terminal backgrounds)
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/CHANGELOG.md]: Removed in a3506b00 — was a leftover from my conflict-resolution of the cherry-pick.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 4]
```diff
@@ -420,6 +423,35 @@ export class Theme {
...
+
+		// Blend away from the terminal background, not unconditionally toward white:
+		// pushing the selection across the background would reduce contrast.
+		const selectionLuminance = luminance(selectionRgb);
+		const top = selectionLuminance >= luminance(terminalBg) ? WHITE : BLACK;
+		const topDelta = Math.abs(luminance(top) - selectionLuminance);
+		const alpha =
+			topDelta > 0
+				? Math.min(SELECTION_MAX_BLEND_ALPHA, (SELECTION_MIN_LUMINANCE_DELTA - delta) / topDelta)
+				: SELECTION_MAX_BLEND_ALPHA;
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in a3506b00 — both blend endpoints are now evaluated: the same-side direction is preferred (within a 1-point tolerance for channel rounding), and if the cap keeps it below the minimum, the blend crosses the background instead. Black-on-near-black now yields a visible highlight.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 5]
```diff
@@ -420,6 +423,59 @@ export class Theme {
...
+			}
+			if (resultDelta > bestDelta) {
+				bestBlend = blended;
+				bestDelta = resultDelta;
+			}
+		}
+		if (!bestBlend) {
+			return (str: string) => this.bg("selectedBg", str);
+		}
+		const adjustedColor = bestAnsiColor(bestBlend, this.mode);
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in a9bfc61f — each candidate blend is now quantized via bestAnsiColor before its delta is evaluated, so 256-color rounding can't erase the contrast. Added a test with exactly this example (#e93e4d on #2ca649): render picks palette 204 (#ff5f87, luminance delta ~28.5) instead of the delta-0 quantization.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 6]
```diff
@@ -420,6 +423,61 @@ export class Theme {
...
+		for (const top of endpoints) {
+			const spread = luminance(top) - selectionLuminance;
+			if (spread === 0) {
+				continue;
+			}
+			const targetLuminance = terminalLuminance + Math.sign(spread) * SELECTION_MIN_LUMINANCE_DELTA;
+			const alpha = Math.min(SELECTION_MAX_BLEND_ALPHA, (targetLuminance - selectionLuminance) / spread);
+			// Quantize before evaluating: on 256-color terminals the palette
+			// rounding can otherwise erase the delta the blend was chosen for.
+			const adjustedColor = bestAnsiColor(blendColor(top, selectionRgb, alpha), this.mode);
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in 0da8c97f — each endpoint now steps the blend from the exact-target alpha up to the 0.5 cap in 0.05 increments, re-quantizing and re-checking at each step, so a stronger blend that quantizes to a passing palette color gets picked. Your example lands on palette 17 (#00005f, delta ~68) via the black endpoint, which is preferred since the selection sits below the terminal background; the alternative white route would quantize similarly well. Added a test for it.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 7]
```diff
@@ -420,6 +423,61 @@ export class Theme {
...
+	getSelectionBackgroundColor(): (str: string) => string {
+		const terminalBg = getDefaultTerminalColors()?.background;
+		const selectionRgb = colorValueToRgb(this.bgColorValues.get("selectedBg"));
+		if (!terminalBg || !selectionRgb) {
+			return (str: string) => this.bg("selectedBg", str);
+		}
+
+		const delta = Math.abs(luminance(selectionRgb) - luminance(terminalBg));
+		if (delta >= SELECTION_MIN_LUMINANCE_DELTA) {
+			return (str: string) => this.bg("selectedBg", str);
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in c6595463 — the baseline delta is now computed from the quantized selection (bestAnsiColor + back-conversion) rather than the configured RGB, so the early return can't skip adaptation when palette rounding erases the contrast.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 8]
```diff
@@ -420,6 +424,66 @@ export class Theme {
 		return this.surfaceBackgroundColor("toolPanelBg");
 	}
 
+	getSelectionBackgroundColor(): (str: string) => string {
+		const terminalBg = getDefaultTerminalColors()?.background;
+		const selectionRgb = colorValueToRgb(this.bgColorValues.get("selectedBg"));
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in c6595463 — basic ANSI indices (0-15) are terminal-defined, so their rendered luminance is unknowable. Adaptation now skips those and returns the raw index.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 9]
```diff
@@ -420,6 +424,66 @@ export class Theme {
...
+				const resultDelta = Math.abs(luminance(adjustedRgb) - terminalLuminance);
+				if (resultDelta >= SELECTION_MIN_LUMINANCE_DELTA - 1) {
+					bestColor = adjustedColor;
+					bestDelta = resultDelta;
+					break;
+				}
+				if (resultDelta > bestDelta) {
+					bestColor = adjustedColor;
+					bestDelta = resultDelta;
+				}
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in c6595463 — bestDelta now starts from the quantized baseline delta, consistent with the quantized candidate measurements.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 10]
```diff
@@ -420,6 +424,66 @@ export class Theme {
...
+			if (spread === 0) {
+				continue;
+			}
+			const targetLuminance = terminalLuminance + Math.sign(spread) * SELECTION_MIN_LUMINANCE_DELTA;
+			const baseAlpha = Math.min(SELECTION_MAX_BLEND_ALPHA, (targetLuminance - selectionLuminance) / spread);
+			// Quantize before evaluating: on 256-color terminals the palette
+			// rounding can otherwise erase the delta the blend was chosen for.
+			// If the direct hit undershoots, keep stepping toward the cap —
+			// a stronger blend may quantize to a palette color that passes.
+			for (let alpha = baseAlpha; alpha <= SELECTION_MAX_BLEND_ALPHA + 1e-9; alpha += SELECTION_BLEND_STEP) {
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in c6595463 — the step sequence now always includes SELECTION_MAX_BLEND_ALPHA as its final value, even when the 0.05 grid starting at baseAlpha wouldn't land on it.

File: packages/coding-agent/src/modes/interactive/theme/theme.ts [hunk 11]
```diff
@@ -420,6 +424,78 @@ export class Theme {
...
 	}
 
+	getSelectionBackgroundColor(): (str: string) => string {
+		const terminalBg = getDefaultTerminalColors()?.background;
+		const selectedBgValue = this.bgColorValues.get("selectedBg");
+		const selectionRgb = colorValueToRgb(selectedBgValue);
+		// Basic ANSI colors (0-15) are terminal-defined; their rendered color is
+		// unknown, so no reliable contrast computation (or blend base) exists.
+		if (!terminalBg || !selectionRgb || (typeof selectedBgValue === "number" && selectedBgValue < 16)) {
+			return (str: string) => this.bg("selectedBg", str);
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/coding-agent/src/modes/interactive/theme/theme.ts]: Fixed in efa39d1c — an empty selectedBg now blends from the detected terminal background itself, so terminal-default selections still get an explicit contrasting highlight (delta 28). Added a test covering it.

**Review by snimu — APPROVED** [review]
